python/tank/platform/software_launcher.py

```diff
@@ -0,0 +1,738 @@
+# Copyright (c) 2016 Shotgun Software Inc.
+#
+# CONFIDENTIAL AND PROPRIETARY
+#
+# This work is provided "AS IS" and subject to the Shotgun Pipeline Toolkit
+# Source Code License included in this distribution package. See LICENSE.
+# By accessing, using, copying or modifying this work you indicate your
+# agreement to the Shotgun Pipeline Toolkit Source Code License. All rights
+# not expressly granted therein are reserved by Shotgun Software Inc.
+
+"""
+Defines the base class for DCC application launchers all Tank engines
+should implement.
+"""
+
+import os
+import sys
+import re
+import logging
+import xml.etree.ElementTree as XML_ET
+
+from ..errors import TankError
+from ..log import LogManager
+from ..util.loader import load_plugin
+
+from . import constants
+from . import validation
+
+from .bundle import resolve_default_value
+from .engine import get_env_and_descriptor_for_engine
+
+# std core level logger
+core_logger = LogManager.get_logger(__name__)
+
+
+def create_engine_launcher(tk, context, engine_name):
+    """
+    Factory method that creates a Toolkit engine specific
+    SoftwareLauncher instance.
+
+    :param tk: :class:`~sgtk.Sgtk` Toolkit instance.
+    :param context: :class:`~sgtk.Context` Context to launch the DCC in.
+    :param engine_name: (str) Name of the Toolkit engine associated with the
+                        DCC(s) to launch.
+    :returns: :class:`SoftwareLauncher` subclass instance.
+    """
+    try:
+        # Get the engine environment and descriptor using Engine.py code
+        (env, engine_descriptor) = get_env_and_descriptor_for_engine(engine_name, tk, context)
+
+        # Make sure it exists locally
+        if not engine_descriptor.exists_local():
+            raise TankError("Cannot create %s software launcher! %s does not exist on disk" %
+                (engine_name, engine_descriptor)
+            )
+
+        # Get path to engine startup code and load it.
+        engine_path = engine_descriptor.get_path()
+        plugin_file = os.path.join(engine_path, constants.ENGINE_SOFTWARE_LAUNCHER_FILE)
+        class_obj = load_plugin(plugin_file, SoftwareLauncher)
+        launcher = class_obj(tk, context, engine_name, env)
+
+    except Exception, e:
+        # Trap and log the exception and let it bubble in unchanged form
+        core_logger.exception("Exception raised in create_engine_launcher :\n%s" % e)
```

Comment:
i don't think we need to log this exception. this might be confusing to a user if it gets displayed in say a shotgun UI. I am wondering if we can just let the exception bubble up and be handled by the caller?

on a side note, `logger.exception()` includes a call stack by default, so i normally just do 

```
except Exception:
    logger.exception("Dragons in the code!")
    raise
```

and the logger takes care of the rest!